### ISSUE TITLE
`dht2` does not work with duikers or amakihi, but does with with sparrows

### ISSUE THREAD

**Issue by erex** [DistanceDevelopment/Distance#73]
Although duiker analysis was [successful on Friday 24Jul](https://github.com/DistanceDevelopment/Distance/files/4971216/howe-vignette-Distance1.0.1.9002.pdf) when reading data from file, now same code does not work (whether reading data from file or from new DuikerCameraTraps object

Reproducible example
```
library(Distance)
data("DuikerCameraTraps")
trunc.list <- list(left=2, right=15)
conversion <- convert_units("meter", NULL, "square kilometer")
peak.hr <- ds(DuikerCameraTraps, transect = "point", key="hr", adjustment = "cos", 
+               cutpoints = c(seq(2,8,1), 10, 12, 15), truncation = trunc.list
viewangle <- 42 # degrees
samfrac <- viewangle / 360
peak.hr.dens <- dht2(peak.hr, flatfile=DuikerCameraTraps, strat_formula = ~1,
+                      sample_fraction = samfrac, er_est = "P2", convert.units = conversion
Error in `$<-.data.frame`(`*tmp*`, "distance", value = numeric(0)) : 
  replacement has 0 rows, data has 6277
```

I can't get a traceback to know any more about where the error occurs.

Furthermore, if I use `dht2` with the amakihi data set, a different error is thrown:
```
data(amakihi)
conv.am <- convert_units("meter", NULL, "hectare")
easy.am <- ds(amakihi, transect="point", key="hr", convert.units = conv.am, truncation="10%")
fraction.am <- dht2(easy.am, flatfile=amakihi, convert_units = conv.am, 
+                     strat_formula = ~Region.Label)
Error in vapply(.x, .f, .mold, ..., USE.NAMES = FALSE) : 
  values must be length 1,
 but FUN(X[[3]]) result is length 2
```

but applying `dht2` to the Savannah_sparrow_1980 data set, there are no problems

```
data("Savannah_sparrow_1980")
sparrow <- ds(Savannah_sparrow_1980, key="hn", transect="point", truncation = "10%",
              convert.units = conv.am)
spar.dht2 <- dht2(sparrow, flatfile = Savannah_sparrow_1980, strat_formula = ~Region.Label,
                  convert_units=conv.am)
print(spar.dht2, report = "density")
```

However, `bootdht` does not seem bothered by the dukier data analysis.  So I can produce estimates of precision, but not point estimates....

**Comment by lenthomas** [comment]
Regarding Duiker, I ran it with ok (having just fetched the newest version eb22ad4 from the master branch at github).  I did have to put a close bracket at the end of the ds() function call, and change convert.units in the dht2 call to convert_units.  Here's the code I used: 
```
data("DuikerCameraTraps")
trunc.list <- list(left = 2, right = 15)
conversion <- convert_units("meter", NULL, "square kilometer")
peak.hr <- ds(DuikerCameraTraps, transect = "point", key = "hr", adjustment = "cos", 
  cutpoints = c(seq(2, 8, 1), 10, 12, 15), truncation = trunc.list)
viewangle <- 42 # degrees
samfrac <- viewangle / 360
peak.hr.dens <- dht2(peak.hr, flatfile = DuikerCameraTraps, strat_formula = ~1,
  sample_fraction = samfrac, er_est = "P2", convert_units = conversion)
print(peak.hr.dens, report = "both")
```
This gave the results
```
Summary statistics:
 .Label  Area CoveredArea   Effort    n  k ER se.ER
  Total 40.37    1015.748 12317058 5865 21  0     0
 cv.ER
 0.302

Abundance estimates:
 .Label Estimate      se    cv LCI  UCI     df
  Total      586 176.963 0.302 316 1085 20.094

Density estimates:
 .Label Estimate    se    cv    LCI     UCI     df
  Total  14.5106 4.384 0.302 7.8351 26.8734 20.094

Component percentages of variance:
 .Label Detection    ER
  Total      0.23 99.77
```

Regarding Amakihi, I get the same error as @erex.

**Comment by erex** [comment]
Thanks @lenthomas . I'll have another bash at the duikers.  I was doing my head in between arguments named `convert.units` (ds) and `convert_units` (dht2) and a function called `convert_units`

**Comment by erex** [comment]
Wow, I've lost the plot.  Cut/paste Prof Thomas' code from comment above, into a clean R environment, with poor results.

```
> data("DuikerCameraTraps")
> trunc.list <- list(left = 2, right = 15)
> conversion <- convert_units("meter", NULL, "square kilometer")
> peak.hr <- ds(DuikerCameraTraps, transect = "point", key = "hr", adjustment = "cos", 
+               cutpoints = c(seq(2, 8, 1), 10, 12, 15), trunc .... [TRUNCATED] 
Starting AIC adjustment term selection.
Fitting hazard-rate key function
Key only model: not constraining for monotonicity.
AIC= 24877.605
Fitting hazard-rate key function with cosine(2) adjustments
AIC= 24879.605

Hazard-rate key function selected.

> viewangle <- 42 # degrees
> samfrac <- viewangle / 360
> peak.hr.dens <- dht2(peak.hr, flatfile = DuikerCameraTraps, strat_formula = ~1,
+                      sample_fraction = samfrac, er_est = "P2", con .... [TRUNCATED] 
Error in `$<-.data.frame`(`*tmp*`, "distance", value = numeric(0)) : 
  replacement has 0 rows, data has 6277
In addition: Warning message:
In dht(model, region.table, sample.table, obs.table, options = dht_options,  :
  Point transect encounter rate variance can only use estimators P2 or P3, switching to P3.
> sessionInfo()
R version 4.0.0 (2020-04-24)
Platform: i386-w64-mingw32/i386 (32-bit)
Running under: Windows 10 x64 (build 19041)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Distance_1.0.1.9004 mrds_2.2.3.9002    
```

**Comment by lenthomas** [comment]
Turns out I lost the plot.  Pilot error on my part meant I was not using Distance_1.0.1.9004.  When I do, I get the same error message as you, @erex.  So, over to you, @dill ...

**Comment by lenthomas** [comment]
In case it's useful: 
```
> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 17763)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252 
[2] LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252
[4] LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods  
[7] base     

other attached packages:
[1] Distance_1.0.1.9004 mrds_2.2.3.9002     testthat_2.3.2     

loaded via a namespace (and not attached):
 [1] pillar_1.4.4        compiler_4.0.2      remotes_2.1.1      
 [4] prettyunits_1.1.1   tools_4.0.2         pkgload_1.1.0      
 [7] digest_0.6.25       pkgbuild_1.0.8      memoise_1.1.0      
[10] lifecycle_0.2.0     tibble_3.0.1        nlme_3.1-148       
[13] lattice_0.20-41     mgcv_1.8-31         pkgconfig_2.0.3    
[16] rlang_0.4.6         Matrix_1.2-18       cli_2.0.2          
[19] rstudioapi_0.11     parallel_4.0.2      withr_2.2.0        
[22] dplyr_1.0.0         fs_1.4.1            desc_1.2.0         
[25] generics_0.0.2      vctrs_0.3.1         devtools_2.3.0     
[28] rprojroot_1.3-2     grid_4.0.2          tidyselect_1.1.0   
[31] glue_1.4.1          R6_2.4.1            processx_3.4.3     
[34] fansi_0.4.1         sessioninfo_1.1.1   Rsolnp_1.16        
[37] purrr_0.3.4         callr_3.4.3         magrittr_1.5       
[40] usethis_1.6.1       backports_1.1.8     ps_1.3.3           
[43] ellipsis_0.3.1      splines_4.0.2       assertthat_0.2.1   
[46] numDeriv_2016.8-1.1 optimx_2020-4.2     truncnorm_1.0-8    
[49] crayon_1.3.4   
```

**Comment by dill** [comment]
Thanks for this folks. The duiker issue should be resolved in 504508c.

The amakihi issue is more concerning as I'm not sure where this bug originated.

@erex: did you use `dht2`+amakihi in your last set of workshops, or was `dht` used for that dataset?

**Comment by erex** [comment]
There is nothing about the amakihi data analysis that requires the use of `dht2` so I have not used that combination in workshops.

But in chasing my tail with the duiker analysis, I sought other point transect data sets to send to `dht2` just to see if the duiker problem was peculiar to duikers or more general.  That's why I turned my attention to amakihi and sparrows.

**Comment by dill** [comment]
Okay, that's fine I was just trying to work out when the bug appeared.

**Comment by dill** [comment]
Okay the issue is actually related to last week's question about `dht`: what to do about `varflag=0` and binomial variance estimation when the study area is the covered area.

Looks like when I wrote [`ER_var_f`](https://github.com/DistanceDevelopment/Distance/blob/master/R/ER_var_f.R#L6) assumed incorrectly that we need to use that estimator when density is estimated (when `Area=0`). The error is thrown when it can't find the `size` column it needs (which is silly given in that case it can't deal with size != 1.

To fix:

1. create new `binomial_var=` argument to `dht2` for this situation. Implement this estimator correctly and `stop` if it's enabled when `size>1`.
2. fix density estimation to have the correct (Innes or otherwise) variance estimator.

**Comment by dill** [comment]
561df96 should fix these issues, though note that the binomial variance estimator is not yet implemented

**Comment by dill** [comment]
(minor follow-up to previous comments that binomial variance was implemented in dba2553e4405ab37c77b8edfdbc5523d9305cc81)
